src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -143,18 +168,36 @@ private boolean commitCursor(final String subscriptionId, final String eventType
 
         final String offsetPath = format(PATH_ZK_OFFSET, subscriptionId, eventType, cursor.getPartition());
         try {
-            return runLocked(() -> {
-                final String currentOffset = new String(zkHolder.get().getData().forPath(offsetPath), CHARSET_UTF8);
-                // Yep, here we are trying to hack a little. This code should be removed during timelines implementation
-                final NakadiCursor first = new NakadiCursor(eventType, cursor.getPartition(), cursor.getOffset());
-                final NakadiCursor second = new NakadiCursor(eventType, cursor.getPartition(), currentOffset);
-                if (topicRepository.compareOffsets(first, second) > 0) {
-                    zkHolder.get().setData().forPath(offsetPath, cursor.getOffset().getBytes(CHARSET_UTF8));
-                    return true;
-                } else {
-                    return false;
-                }
-            }, zkLockFactory.createLock(offsetPath));
+            @SuppressWarnings("unchecked")
+            final Boolean committed = executeWithRetry(() -> {
+                        final Stat stat = new Stat();
+                        final byte[] currentOffsetData = zkHolder.get()
+                                .getData()
+                                .storingStatIn(stat)
+                                .forPath(offsetPath);
+                        final String currentOffset = new String(currentOffsetData, Charsets.UTF_8);
+
+                        // Yep, here we are trying to hack a little. This code
+                        // should be removed during timelines implementation
+                        final NakadiCursor cursorToCommit = new NakadiCursor(eventType, cursor.getPartition(),
+                                cursor.getOffset());
+                        final NakadiCursor currentCursor = new NakadiCursor(eventType, cursor.getPartition(),
+                                currentOffset);
+                        if (topicRepository.compareOffsets(cursorToCommit, currentCursor) > 0) {
+                            zkHolder.get()
+                                    .setData()
+                                    .withVersion(stat.getVersion())
+                                    .forPath(offsetPath, cursor.getOffset().getBytes(Charsets.UTF_8));
+                            return true;
+                        } else {
+                            return false;
+                        }
+                    },
+                    new RetryForSpecifiedCountStrategy<Boolean>(COMMIT_CONFLICT_RETRY_TIMES)
+                            .withExceptionsThatForceRetry(KeeperException.BadVersionException.class));
+
+            return Optional.ofNullable(committed).orElse(false);
```

Comment:
executeWithRetry is `@Nullable` so I don't really know if it will return null or not.